docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,803 @@
+removing realm vlobs (cannot remove a realm given it is related to certificates) or
+organization form the database.
+
+### 5.3 Server data migration
+
+Drop tables:
+
+- `vlob_encryption_revision`
+- `message`
+
+Table `vlob_atom`:
+
+- Remove all rows that contains old encryption revision data (i.e. the table only contains
+  the vlob currently accessible by the users).
+- Remove column `vlob_encryption_revision`
+- Add column `key_index` (`INTEGER NOT NULL`). This defaults to 1 when migrating data.
+
+Table `realm`
+
+- Remove columns `encryption_revision`, `maintenance_started_by`, `maintenance_started_on` and `maintenance_type`
+
+## 6 Changes: Client-side
+
+### 6.1 Certificate Ops
+
+`RealmKeyRotationCertificate` should be checked (only OWNER of the realm can create such
+certificate).
+
+### 6.2 Fetching keys bundle
+
+The client lazily fetch the realm keys bundle: it do the fetch operation when a server
+operation requires it (i.e. `realm_share`, `vlob_read`, `vlob_create` and `Vlob_update`).
+
+The realm keys bundle is only kept in RAM, not on disk.
+
+When fetching keys bundle and keys bundle access, both must be checked against the
+corresponding `RealmKeyRotationCertificate` (same author and timestamp).
+
+At any time, `realm_share`/`vlob_read`/`vlob_create` can return `bad_key_index` if the client
+consider the wrong key to be the last one. In this case the client simply has to forget
+it realm keys bundle and re-do the fetch operation.
+
+In case the current keys bundle is corrupted (e.g. bad signature, missing keys etc.),
+a self-healing is attempted:
+
+1. client looks into the certificates to find previous key rotation certificates
+2. client calls `realm_get_keys_bundle` with a `timestamp` parameter set to the key
+   rotation certificate.
+
+   - If keys bundle is corrupted, go to 1. until there is no more keys bundles
+   - If keys bundle can be retrieve, the self-healing has succeeded.
+   - If the keys bundle was never available to our user (i.e. wasn't the last one when
+     the realm got shared to us), the self-heading has failed.
+
+3. Client dispatch internally an event to inform of corruption (including the author
+   of the faulty key rotation) and self-healing outcome. This event typically ends
+   up as a warning logged and gets displayed in the GUI.
+
+### 6.3 Datamodel
+
+> TODO
+
+### 6.4 Automatic key rotation
+
+For automatic key rotation, a new "key rotation manager" is to be introduced. It role
+is very simple:
+
+1. Wait for new `UserRevoked`, `RealmRoleUpdate` and `RealmKeyRotationCertificate` certificates
+2. For `UserRevoked`, call `realm_unshare` for each realm it is owner (and the revoked
+  user is part of)
+3. For `RealmRoleUpdate`, if the role is None and the user is owner of the realm:
+  3.1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
+       already waiting for this realm)
+  3.2. Call `realm_rotate_key`
```

Comment:
```suggestions
  a. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  b. Call `realm_rotate_key`
```

Reply:
I think it's better to keep number instead of letter here given 3.1 and 3.2 are ordered (and not two separated possibilities)

Reply:
We can keep the number, it's more about the numbering format used that isn't supported by markdown (it wont be presented as a sub-list)

```suggestion
  1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  2. Call `realm_rotate_key`
``` 

or 

```suggestion
  1) Wait a random time between 30 and 60 seconds (and reset the wait if we were
       already waiting for this realm)
  2) Call `realm_rotate_key`
```